File: agent/contracts/deploy.py
"""Deploys VeriFiEscrow to Arc testnet using the `escrow` role's wallet as
deployer/owner, with the `orchestrator` role's wallet set as the settler
(the only address allowed to release/finalize/refund a job -- matches who
already executes settlement/escrow.py's settle() today).

Writes the deployed address to contracts/deployed_address.txt so
payments/escrow_contract.py can pick it up without redeploying every run.

Run standalone:
    python -m contracts.deploy
"""
from __future__ import annotations
import json
import os

from web3 import Web3

from shared.config import ARC_RPC_URL, ARC_CHAIN_ID, WALLETS
from contracts.compile import compile_contract, ARTIFACT_PATH, CONTRACTS_DIR

DEPLOYED_ADDRESS_PATH = os.path.join(CONTRACTS_DIR, "deployed_address.txt")


def deploy() -> str:
    if os.path.exists(ARTIFACT_PATH):
        with open(ARTIFACT_PATH) as f:
            artifact = json.load(f)
    else:
        artifact = compile_contract()

    w3 = Web3(Web3.HTTPProvider(ARC_RPC_URL))
    deployer = WALLETS["escrow"]
    settler_address = WALLETS["orchestrator"]["address"]

    contract = w3.eth.contract(abi=artifact["abi"], bytecode=artifact["bytecode"])
    tx = contract.constructor(Web3.to_checksum_address(settler_address)).build_transaction({
        "from": Web3.to_checksum_address(deployer["address"]),
        "nonce": w3.eth.get_transaction_count(Web3.to_checksum_address(deployer["address"]), "pending"),
        "gas": 1_500_000,
        "gasPrice": w3.eth.gas_price,
        "chainId": ARC_CHAIN_ID,
    })
    signed = w3.eth.account.sign_transaction(tx, deployer["private_key"])
    tx_hash = w3.eth.send_raw_transaction(signed.raw_transaction)
    print(f"deploy tx: {tx_hash.hex()} -- waiting for receipt...")
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash, timeout=90)
    if receipt["status"] != 1:
        raise RuntimeError(f"deployment reverted: {tx_hash.hex()}")

    address = receipt["contractAddress"]
    print(f"VeriFiEscrow deployed at {address} (block {receipt['blockNumber']}, settler={settler_address})")

    with open(DEPLOYED_ADDRESS_PATH, "w") as f:
        f.write(address)
    return address


if __name__ == "__main__":
    deploy()

File: agent-ts/packages/contracts/deploy.ts
/**
 * Deploys VeriFiEscrow to Arc testnet using the `escrow` role's wallet as
 * deployer/owner, with the `orchestrator` role's wallet set as the settler
 * (the only address allowed to release/finalize/refund a job -- matches
 * who already executes settlement.ts's settle() today). Ported from
 * agent/contracts/deploy.py.
 *
 * Reuses the ABI + bytecode already compiled for the Python version
 * (VeriFiEscrow.json, copied alongside this file) -- no recompile needed,
 * a contract's bytecode doesn't care what language deployed it.
 *
 * Writes the deployed address to deployed_address.txt so escrowContract.ts
 * can pick it up without redeploying every run.
 *
 * Run standalone:
 *   npx tsx packages/contracts/deploy.ts
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { createPublicClient, createWalletClient, http, getAddress } from "viem";
import { privateKeyToAccount } from "viem/accounts";

import { config, chain } from "@arcproof/core";

const { ARC_RPC_URL, WALLETS } = config;
const { arcTestnet } = chain;

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ARTIFACT_PATH = path.join(__dirname, "VeriFiEscrow.json");
const DEPLOYED_ADDRESS_PATH = path.join(__dirname, "deployed_address.txt");

async function deploy(): Promise<string> {
  const artifact = JSON.parse(fs.readFileSync(ARTIFACT_PATH, "utf-8"));
  const deployerWallet = WALLETS.escrow;
  const settlerAddress = WALLETS.orchestrator.address;

  if (!deployerWallet.privateKey) throw new Error("no private key configured for role 'escrow' -- check .env (run npm run gen-wallets first)");
  if (!settlerAddress) throw new Error("no address configured for role 'orchestrator' -- check .env (run npm run gen-wallets first)");

  const account = privateKeyToAccount(deployerWallet.privateKey as `0x${string}`);
  const publicClient = createPublicClient({ chain: arcTestnet, transport: http(ARC_RPC_URL) });
  const walletClient = createWalletClient({ account, chain: arcTestnet, transport: http(ARC_RPC_URL) });

  const txHash = await walletClient.deployContract({
    abi: artifact.abi,
    bytecode: artifact.bytecode as `0x${string}`,
    args: [getAddress(settlerAddress)],
  });
  console.log(`deploy tx: ${txHash} -- waiting for receipt...`);
  const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash, timeout: 90_000 });
  if (receipt.status !== "success" || !receipt.contractAddress) {
    throw new Error(`deployment reverted: ${txHash}`);
  }

  const address = receipt.contractAddress;
  console.log(`VeriFiEscrow deployed at ${address} (block ${receipt.blockNumber}, settler=${settlerAddress})`);

  fs.writeFileSync(DEPLOYED_ADDRESS_PATH, address);
  console.log(`\nSave this into agent-ts/.env: ESCROW_CONTRACT_ADDRESS=${address}`);
  console.log(`(also mirrored at ${DEPLOYED_ADDRESS_PATH} for escrowContract.ts to read directly)`);
  return address;
}

deploy().catch((e) => {
  console.error(e);
  process.exit(1);
});
